[RFC] edtlib: enhance for pin controller, clock controller, partitions, special nodes

b0661 opened the pull request:
A set of edtlib.py enhancements that proved useful to implement an OOT pin controller driver, clock controller driver, flash driver, gpio leds driver (using code generation). 

This makes edtlib more complete. It should also simplify defines generation in gen_defines.py, especially for pin controllers and clock controllers. 

Comments welcome.

dts: edtlib: enable bindings for special nodes without compatibles
--------------------------------------------------------------------------------------
```
Associate a default compatible to special nodes that by specification
do not have or can not have a compatible.
The default compatible allows to use bindings for nodes that otherwise
can not be controlled by a binding.

Default compatible for special DTS nodes:
  - '/aliases' node                       : 'aliases'
  - '/chosen' node                        : 'chosen'
  - pin controller pinctrl state node     : 'pinctrl-state'
  - pin controller pin configuration node : 'pincfg'
  - fixed partition node                    : 'fixed-partition'
  - gpio led node                              : 'gpio-led'

Default compatibles are only associated if no compatible is given.
```

dts: edtlib: allow binding override by binding file name
-----------------------------------------------------------------------
```
Allow to override a binding by file name. Take the first file
given by the binding directories provided to edtlib.

This allows applications and out of tree drivers to amend
a binding already given by Zephyr.

Warn in case there are duplicate binding file names.
```

dts: edtlib: allow extraction of gpio-ranges properties
---------------------------------------------------------------------
Handling of 'gpio-ranges' is a must to implement pin controllers.  
```
gpio-ranges properties expect a phandle value list.
Extend _standard_phandle_val_list() to also work on gpio-ranges.
```

dts: edtlib: extract pinctrl state and pin configuration nodes
-----------------------------------------------------------------------------
```
Extract the pin controller child nodes for pinctrl state and pin
configuration.

Extract gpio ranges controlled by the pin controller.

Provide a PinCfg object that represents the pin configuration
data for easy access.

Extend the Node class with properties for:

  pinctrl_states:
    A list with the Node instances for the 'pinctrl-state' children of
    a pin controller node. The list is empty if the node does not have any
    pinctrl state children.

  pincfgs:
    A list of PinCfg objects for the 'pinctrl-state' node. The list is
    empty if the node is not a 'pinctrl-state' node.

  pinctrl_gpio_ranges:
    A list of ControllerAndData objects of the gpio ranges a pin controller
    controls. The list is empty if the node is not a pin controller or no
    'gpio-ranges' are defined by the gpio nodes.

  pin_controller:
    The pin controller for the node. Only meaningful for nodes representing
    pinctrl states.
```

dts: edtlib: extract clock inputs and clock outputs
----------------------------------------------------------------
```
Extract clock inputs ('clocks') and clock outputs ('clock-output-names')
properties.

Extend the Node class with properties for:

  clocks:
    A list of ControllerAndData objects for the clock inputs on the
    node, sorted by index. The list is empty if the node does not have a
    clocks property.

  clock_outputs:
    A list of ControllerAndData objects for the clock outputs on the
    node, sorted by index. The list is empty if the node does not have a
    \#clock-cells property.
```

dts: edtlib: extract partitions nodes
---------------------------------------------
```
Extract the partitions child node of flash nodes.

Extend the Node class with properties for:

  partitions:
    A list of Partition objects of the partitions of the 'partitions'
    node of a flash. Only meaningful for nodes representing a flash -
    otherwise an empty list.
```
dts: edtlib: extract gpio led nodes
-----------------------------------------

```
Extract the gpio led child node of gpio leds controller nodes.

Extend the Node class with properties for:

  gpio_leds:
    A list of ControllerAndData objects of the leds a gpio leds
    controller controls. The list is empty if the node is not a
    gpio leds controller or it does not have gpio led children.
```

dts: edtlib: issue same warning only once
-------------------------------------------------------
Repeating the same warning is just an annoyance.
```
Issue a warning only on first occurence. Suppress all following
warnings having the identical warning text.
```

b0661 committed:
dts: edtlib: issue same warning only once

Issue a warning only on first occurence. Suppress all following
warnings having the identical warning text.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 committed:
dts: edtlib: allow extraction of gpio-ranges properties

gpio-ranges properties expect a phandle value list.
Extend _standard_phandle_val_list() to also work on gpio-ranges.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 committed:
dts: edtlib: enable bindings for special nodes without compatibles

Associate a default compatible to special nodes that by specification
do not have or can not have a compatible.
The default compatible allows to use bindings for nodes that otherwise
can not be controlled by a binding.

Default compatible for special DTS nodes:
  - '/aliases' node                       : 'aliases'
  - '/chosen' node                        : 'chosen'
  - pin controller pinctrl state node     : 'pinctrl-state'
  - pin controller pin configuration node : 'pincfg'
  - fixed partion node                    : 'fixed-partition'
  - gpio led node                         : 'gpio-led'

Default compatibles are only associated if no compatible is given.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 committed:
dts: edtlib: allow binding override by binding file name

Allow to override a binding by file name. Take the first file
given by the binding directories provided to edtlib.

This allows applications and out of tree drivers to amend
a binding already given by Zephyr.

Warn in case there are duplicate binding file names.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 committed:
dts: edtlib: extract pinctrl state and pin configuration nodes

Extract the pin controller child nodes for pinctrl state and pin
configuration.

Extract gpio ranges controlled by the pin controller.

Provide a PinCfg object that represents the pin configuration
data for easy access.

Extend the Node class with properties for:

  pinctrl_states:
    A list with the Node instances for the 'pinctrl-state' children of
    a pin controller node. The list is empty if the node does not have
    any pinctrl state children.

  pincfgs:
    A list of PinCfg objects for the 'pinctrl-state' node. The list is
    empty if the node is not a 'pinctrl-state' node.

  pinctrl_gpio_ranges:
    A list of ControllerAndData objects of the gpio ranges a pin
    controller controls. The list is empty if the node is not a pin
    controller or no 'gpio-ranges' are defined by the gpio nodes.

  pin_controller:
    The pin controller for the node. Only meaningful for nodes
    representing pinctrl states.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 committed:
dts: edtlib: extract clock inputs and clock outputs

Extract clock inputs ('clocks') and clock outputs ('clock-output-names')
properties.

Extend the Node class with properties for:

  clocks:
    A list of ControllerAndData objects for the clock inputs on the
    node, sorted by index. The list is empty if the node does not have a
    clocks property.

  clock_outputs:
    A list of ControllerAndData objects for the clock outputs on the
    node, sorted by index. The list is empty if the node does not have a
    \#clock-cells property.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 committed:
dts: edtlib: extract partitions nodes

Extract the partitions child node of flash nodes.

Extend the Node class with properties for:

  partitions:
    A list of Partition objects of the partitions of the 'partitions'
    node of a flash. Only meaningful for nodes representing a flash -
    otherwise an empty list.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -1144,6 +1224,107 @@ def _check_undeclared_props(self):
+                prop_value.controller.pinctrl_gpio_ranges = []
+            prop_value.controller.pinctrl_gpio_ranges.append(entry)
+
+    def _init_clock_outputs(self):
```
Quick look, but this function looks pretty long and hard to follow.

Could you try splitting it up into some smaller functions, so that each functions is like 30 lines maximum? Maybe some common stuff could be factored out at the same time.

b0661 replied:
Split. Did not reach 30 lines. Where do get this limit from?

ulfalizer replied:
Just some number I pulled out that feels like the right ballpark. Short and focused functions with few parameters are considered a good practice at least. See e.g. the [Linux kernel coding style](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#functions).

Alright to have it a bit longer if it's just a big switch or whatever with only one or two levels of indent, because then it's more like many small functions.

Good idea to go through [PEP 8](https://www.python.org/dev/peps/pep-0008/) once too. Check [this section](https://www.python.org/dev/peps/pep-0008/#indentation) for how to format continuation lines (e.g. for function parameters).

Strive for having few parameters too, and for all combinations of parameters to make sense (e.g., no dependencies between parameters, if it can be avoided).

Might be a bit busy next week, but will look more at the logic.

ulfalizer replied:
Not saying I'm perfect re. that stuff. Think it might be possible to shorten some of this stuff though, and have some simple helpers with just a few parameters.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -1173,6 +1354,142 @@ def _init_regs(self):
+
+        self.partitions = _partitions_list(self.children['partitions'], self)
+
+    def _init_pinctrl_state(self, pinctrl_node, pinctrl_state_node):
```
Ditto here. See if it can be split up a bit.

b0661 replied:
Split. Did not reach 30 lines.

ulfalizer commented:
Will need to add tests in `testedtlib.py` too.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -1144,6 +1224,107 @@ def _check_undeclared_props(self):
+        clock_cells = node.props["#clock-cells"].to_num()
+        clock_output_names = node.props.get("clock-output-names", [])
+        if clock_output_names:
+            clock_output_names = clock_output_names.to_strings()
```
Could make it a bit easier to read like this:

```python
if "clock-output-names" in node.props:
    clock_output_names = node.props["clock-output-names"].to_strings()
else:
    clock_output_names = []
```

I try to avoid the "set a variable to a value that might be wrong, do a check, and then overwrite it with the right value" pattern. Gets confusing when scanning for where a variable is set.

b0661 replied:
Changed programming style to be in line.

ulfalizer commented:
Improved the CI script so that it still runs other tests when there's an internal error in `test(e)dtlib.py`: https://github.com/zephyrproject-rtos/ci-tools/pull/127. Sorry about that.

There's these issues reported by pylint in the meantime:

```
ERROR   : Test pylint failed: ************* Module edtlib
scripts/dts/edtlib.py:1571:0: W0311: Bad indentation. Found 16 spaces, expected 12 (bad-indentation)
scripts/dts/edtlib.py:1572:0: W0311: Bad indentation. Found 16 spaces, expected 12 (bad-indentation)
scripts/dts/edtlib.py:1574:0: W0311: Bad indentation. Found 16 spaces, expected 12 (bad-indentation)
scripts/dts/edtlib.py:1575:0: W0311: Bad indentation. Found 16 spaces, expected 12 (bad-indentation)
scripts/dts/edtlib.py:1576:0: W0311: Bad indentation. Found 16 spaces, expected 12 (bad-indentation)
scripts/dts/edtlib.py:1577:0: W0311: Bad indentation. Found 16 spaces, expected 12 (bad-indentation)
scripts/dts/edtlib.py:1578:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1580:0: W0311: Bad indentation. Found 24 spaces, expected 20 (bad-indentation)
scripts/dts/edtlib.py:1581:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1582:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1584:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1585:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1586:0: W0311: Bad indentation. Found 24 spaces, expected 20 (bad-indentation)
scripts/dts/edtlib.py:1588:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1589:0: W0311: Bad indentation. Found 24 spaces, expected 20 (bad-indentation)
scripts/dts/edtlib.py:1591:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1592:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1594:0: W0311: Bad indentation. Found 16 spaces, expected 12 (bad-indentation)
scripts/dts/edtlib.py:1595:0: W0311: Bad indentation. Found 16 spaces, expected 12 (bad-indentation)
scripts/dts/edtlib.py:1596:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1600:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1601:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1602:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1603:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1604:0: W0311: Bad indentation. Found 20 spaces, expected 16 (bad-indentation)
scripts/dts/edtlib.py:1606:0: W0311: Bad indentation. Found 16 spaces, expected 12 (bad-indentation)
scripts/dts/edtlib.py:1608:0: W0311: Bad indentation. Found 16 spaces, expected 12 (bad-indentation)
scripts/dts/edtlib.py:1362:11: C0113: Consider changing "not 'partitions' in self.children" to "'partitions' not in self.children" (unneeded-not)
scripts/dts/edtlib.py:2547:8: W0612: Unused variable 'partition_name' (unused-variable)
```

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -2085,6 +2540,33 @@ def _phandle_val_list(prop, n_cells_name):
+            partition.data[prop_name] = prop.val
+        if 'partitions' in partition_node.children:
+            child_partitions = partition_node.children['partitions']
+            partition.data['partitions'] = _partitions_list(child_partitions,
```
I think we should avoid magic hardcoded strings like these that "piggyback" on dictionaries used for other stuff. Better to use regular attributes.

Maybe a new class could be added.

b0661 replied:
New class 'Partition' used.

ulfalizer commented on scripts/dts/edtlib.py:
```diff
@@ -2164,6 +2646,12 @@ def _check_dt(dt):
 def _err(msg):
     raise EDTError(msg)
 
+_warnings = list()
```
Better to have this as part of the `EDT` class, since it's possible to have multiple `EDT` instances. The library itself has no global state.

b0661 replied:
Moved to EDT. Made some collateral moves necessary as _warn now needs EDT.

b0661 commented:
Sorry made a push of unrelated material. This added a lot of unrelated reviewers.

b0661 committed:
dts: edtlib: test pin and clock controller, partitions, special nodes

Tests for edtlib features:
- enable bindings for special nodes without compatibles
- allow binding override by binding file name
- allow extraction of gpio-ranges properties
- extract pinctrl state and pin configuration nodes
- extract clock inputs and clock outputs
- extract partitions nodes
- extract gpio led nodes

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 committed:
dts: edtlib: extract gpio led nodes

Extract the gpio led child node of gpio leds controller nodes.

Extend the Node class with properties for:

  gpio_leds:
    A list of ControllerAndData objects of the leds a gpio leds
    controller controls. The list is empty if the node is not a
    gpio leds controller or it does not have gpio led children.

Signed-off-by: Bobby Noelte <b0661n0e17e@gmail.com>

b0661 commented:
Further expanded edtlib.py to provide a list of ControllerAndData objects of the leds controlled by a gpio leds controller. Also the led node can now have a binding. This is just another application of the  "dts: edtlib: enable bindings for special nodes without compatibles" mechanism.
The extension was created because our new board happens to have a 3color led. The new function gpio_leds() could also support #21578 in the generation of macros for the gpio led driver.

b0661 commented:
@ulfalizer, thank you for reviewing.

The PR should be pretty complete now. There is a false negative Gitlint issue that relates to a commit that is not on the PR branch. I don't know how this can happen.

I'm  unsure how detailed the testing bindings shall be. I provided detailed bindings for the special nodes so they can be used as a sample for real usage. The other testing bindings are kept to a minimum.

> Might be a bit busy next week, but will look more at the logic.

Thank you

b0661 commented:
@ulfalizer, any further comments? What to do to get this merged?

mbolivar commented:
> @ulfalizer, any further comments? What to do to get this merged?

He is currently on vacation

b0661 commented:
Stale too long
